File: 2021/typescript/src/days/day1.ts
import {Day} from "../aoc";
import {sum} from "../aoc";

export class Day1 extends Day {
    day = (): number => 1;
    private depth: number[];

    part1 = () => this.increases(1);
    part2 = () => this.increases(3);
    setup = () => this.depth = this.input.map(Number);

    private increases = (ws = 1) => this.depth
        .filter(( cur, i, l) => i !== 0 && i <= l.length - ws && this.window(i, ws) > this.window(i - 1, ws))
        .length;

    private window = (i: number, size: number): number => this.depth.slice(i, i + size).reduce(sum);
}

File: 2021/julia/src/Day1.jl
module Day1

include("Aoc.jl")
using .Aoc

depth = parse.(Int64, Aoc.input_lines(1))
part1() = count(((i, v),) -> i > 1 && v > depth[i-1], enumerate(depth))
part2() = count(((i, v),) -> i > 3 && sum(depth[i-2:i]) > sum(depth[i-3:i-1]), enumerate(depth))

end